File: library/src/main/java/org/xmtp/android/library/messages/Topic.kt
package org.xmtp.android.library.messages

sealed class Topic {
    data class userWelcome(val installationId: String?) : Topic()
    data class groupMessage(val groupId: String?) : Topic()

    val description: String
        get() {
            return when (this) {
                is groupMessage -> wrapMls("g-$groupId")
                is userWelcome -> wrapMls("w-$installationId")
            }
        }

    private fun wrapMls(value: String): String = "/xmtp/mls/1/$value/proto"
}

File: Sources/XMTPiOS/Topic.swift
//
//  Topic.swift
//
//
//  Created by Pat Nakajima on 11/17/22.
//

public enum Topic {
	case userWelcome(String),
		 groupMessage(String)

	var description: String {
		switch self {
		case let .groupMessage(groupId):
			return wrapMls("g-\(groupId)")
		case let .userWelcome(installationId):
			return wrapMls("w-\(installationId)")
		}
	}
	
	private func wrapMls(_ value: String) -> String {
		"/xmtp/mls/1/\(value)/proto"
	}

    static func isValidTopic(topic: String) -> Bool {
        return topic.allSatisfy(\.isASCII)
    }
}
